src/librustc_mir/interpret/validity.rs

```diff
@@ -95,126 +128,189 @@ fn path_format(path: &Vec<PathElem>) -> String {
+    match value {
+        ScalarMaybeUndef::Undef =>
+            "uninitialized bytes".to_owned(),
+        ScalarMaybeUndef::Scalar(Scalar::Ptr(_)) =>
+            "a pointer".to_owned(),
+        ScalarMaybeUndef::Scalar(Scalar::Bits { bits, .. }) =>
+            bits.to_string(),
+    }
+}
+
 impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M> {
-    fn validate_scalar(
+    /// Make sure that `value` is valid for `ty`
+    fn validate_scalar_type(
         &self,
         value: ScalarMaybeUndef,
         size: Size,
-        scalar: &layout::Scalar,
         path: &Vec<PathElem>,
         ty: Ty,
+        const_mode: bool,
     ) -> EvalResult<'tcx> {
-        trace!("validate scalar: {:#?}, {:#?}, {:#?}, {}", value, size, scalar, ty);
-        let (lo, hi) = scalar.valid_range.clone().into_inner();
+        trace!("validate scalar by type: {:#?}, {:#?}, {}", value, size, ty);
 
-        let value = match value {
-            ScalarMaybeUndef::Scalar(scalar) => scalar,
-            ScalarMaybeUndef::Undef => return validation_failure!("undefined bytes", path),
-        };
+        // Go over all the primitive types
+        match ty.sty {
+            ty::Bool => {
+                try_validation!(value.to_bool(),
+                    scalar_format(value), path, "a boolean");
+            },
+            ty::Char => {
+                try_validation!(value.to_char(),
+                    scalar_format(value), path, "a valid unicode codepoint");
+            },
+            ty::Float(_) | ty::Int(_) | ty::Uint(_) if const_mode => {
+                // Integers/floats in CTFE: Must be scalar bits
+                try_validation!(value.to_bits(size),
+                    scalar_format(value), path, "initialized plain bits");
+            }
+            ty::Float(_) | ty::Int(_) | ty::Uint(_) | ty::RawPtr(_) => {
+                // Anything but undef goes
+                try_validation!(value.not_undef(),
+                    scalar_format(value), path, "a raw pointer");
+            },
+            ty::Ref(..) => {
+                // This is checked by the recursive reference handling, nothing to do here.
+                debug_assert!(ty.builtin_deref(true).is_some() && !ty.is_unsafe_ptr());
+            }
+            ty::FnPtr(_sig) => {
+                let ptr = try_validation!(value.to_ptr(),
+                    scalar_format(value), path, "a pointer");
+                let _fn = try_validation!(self.memory.get_fn(ptr),
+                    scalar_format(value), path, "a function pointer");
+                // TODO: Check if the signature matches
+            }
+            ty::FnDef(..) => {
+                // This is a zero-sized type with all relevant data sitting in the type.
+                // There is nothing to validate.
+            }
+            _ => bug!("Unexpected primitive type {}", ty)
+        }
+        Ok(())
+    }
 
+    /// Make sure that `value` matches the
```

Comment:
"the....."

Reply:
I see `value` as the subject in this sentence, so it shouldn't need a "the". It's like a name: "Make sure that Berlin matches..."